Added telemetry to DevServer

carldebilly opened the pull request:
Closes https://github.com/unoplatform/uno-private/issues/1373


# ✨ Feature

This PR adds comprehensive telemetry support to the DevServer and related components, updates dependencies for telemetry, and introduces end-to-end tests validating telemetry flows.

- Integrate Uno.DevTools.Telemetry across server, host, and discovery pipelines
- Update project files to register telemetry services and adapt SLN for new test projects
- Add extensive MSTest test suites for telemetry, scoped services, and helper classes

## What is the new behavior? 🚀
This pull request introduces telemetry enhancements, dependency updates, and new tests for the Uno DevServer. The changes focus on improving telemetry data handling, adding new test coverage, and updating project dependencies to support testing and telemetry features.

### Telemetry Enhancements

* Added a detailed telemetry events inventory to `Telemetry/Telemetry.md`, documenting event properties, measurements, and anonymization practices for GDPR/privacy compliance.

### Dependency Updates

* Updated `global.json` to include the `MSTest.Sdk` dependency (version 3.8.3) for enhanced testing capabilities.

### New Test Coverage

#### DevServer Tests
* Implemented multiple tests in `DevServerTests.cs` to validate DevServer functionality, including startup, output capture, and clean shutdown.
* Added global usings for common libraries (`FluentAssertions`, `Microsoft.Extensions.Logging`, etc.) in `GlobalUsings.cs` to simplify test code.

#### Telemetry Tests
* Created `TelemetryHashHelperTests.cs` to ensure deterministic and distinct hashing for telemetry data, handling edge cases like null and empty inputs.
* Added `Telemetry/ContextualTelemetryTests.cs` to validate telemetry behavior in global-only scenarios.
* Added `Telemetry/ServerTelemetryTests.cs` to ensure server logs connection events correctly during telemetry cycles.

#### Scoped Service Isolation
* Developed `ScopedServiceIsolationTests.cs` to verify scoped service isolation between connections and proper telemetry redirection.

#### Helper Classes
* Added `SolutionHelper.cs` to facilitate temporary solution file creation for testing purposes.


## PR Checklist ✅

Please check if your PR fulfills the following requirements:

- [X] 📝 Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.
- [X] 🧪 Added [Runtime tests, UI tests, or a manual test sample](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] 📚 Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] 🖼️ Validated PR `Screenshots Compare Test Run` results.
- [X] ❗ Contains **NO** breaking changes

carldebilly committed:
feat(telemetry): implement telemetry for hot reload events

carldebilly committed:
feat(test): add DevServer test infrastructure with telemetry redirection support

carldebilly committed:
feat: enhance devserver testing infrastructure and implement lifecycle telemetry

carldebilly committed:
feat: implement CTRL-C via stdin for graceful devserver shutdown

carldebilly committed:
feat(telemetry): implement server-side telemetry logging - add-in discovery, loading, and processor discovery with comprehensive error handling and timing measurements

carldebilly committed:
test(telemetry): add advanced integration and isolation tests for analytics and connection context

- Update telemetry, DI, and context logic for per-connection analytics
- Add ContextualTelemetryTests for global/connection context and event isolation
- Add ScopedServiceIsolationTests for scoped service and telemetry isolation
- Add TelemetryTestBase for test utilities
- Improve file telemetry naming and session metadata
- Refactor and document advanced analytics architecture

carldebilly committed:
chore: Fixed broken compilation problem in non-devserver projects (code was leaking)

carldebilly committed:
refactor: Modernize and unify DevServer telemetry tests

- Add telemetry events inventory for GDPR/privacy review (Telemetry/Telemetry.md)
- Centralize telemetry test logic in TelemetryTestBase with unified helpers
- Refactor all telemetry tests to use consistent patterns and naming
- Add comprehensive DevServerTestHelper for robust test server management
- Replace obsolete/redundant tests with focused, maintainable test classes
- Implement single-file telemetry output with context prefixes (global/connection-*)
- Update project configuration for proper test dependencies and global usings
- Enhance AddIns discovery with proper telemetry integration and privacy-aware properties
- Fix service registration and DI patterns for better testability
- Update documentation (todos.md, R-D.md) to reflect completed work

Test coverage: 7/7 tests passing with improved reliability and diagnostics

carldebilly committed:
feat: implement centralized telemetry anonymization with MD5 hashing

- Add TelemetryHashHelper with high-performance MD5 implementation
- Fix SolutionId generation to use hash(solutionPath + machineName)
- Anonymize sensitive fields: SolutionId, MachineName, RemoteIpAddress, BasePath, UserAgent
- Restructure tests into Telemetry/ and Hashing/ directories with Given/When/Then naming
- Update documentation to clarify anonymized vs raw telemetry fields
- Link TelemetryHashHelper across test projects for comprehensive validation

carldebilly committed:
perf: migrate duration measurements from DateTime to Stopwatch API

- Replace DateTime.UtcNow with Stopwatch.GetTimestamp()/GetElapsedTime() for all timing measurements
- Improve telemetry precision and performance across DevServer components
- Preserve actual timestamps (CreatedAt, ConnectedAt) that require real DateTime values
- Apply migration to startup/shutdown timing, add-in discovery, assembly loading, and processor discovery
- Update test helpers for consistent timing methodology

carldebilly committed:
feat(hotreload-telemetry): Add server-side Hot Reload telemetry measurements and documentation

- Add FileCount and DurationMs metrics to all Hot Reload events in ServerHotReloadProcessor.Notify
- Document all Hot Reload telemetry events, properties, and measurements in Telemetry.md

unodevops commented:
🤖 Your WebAssembly Skia Sample App stage site is ready! Visit it here: https://unowasmprstaging.z20.web.core.windows.net/pr-20951/wasm-skia-net9/index.html

carldebilly committed:
ci: include Uno.UI.RemoteControl.DevServer.Tests in unit test solution filter for CI

Ensures DevServer server-side tests are built and run in Azure DevOps pipelines.

carldebilly committed:
ci: specify MSTest.Sdk version 3.8.3 for DevServer test project

Fixes CI build error by explicitly setting the MSTest.Sdk version in Uno.UI.RemoteControl.DevServer.Tests.csproj.

unodevops commented:
🤖 Your WebAssembly Skia Sample App stage site is ready! Visit it here: https://unowasmprstaging.z20.web.core.windows.net/pr-20951/wasm-skia-net9/index.html

unodevops commented:
🤖 Your WebAssembly Skia Sample App stage site is ready! Visit it here: https://unowasmprstaging.z20.web.core.windows.net/pr-20951/wasm-skia-net9/index.html

unodevops commented:
🤖 Your Docs stage site is ready! Visit it here: https://unodocsprstaging.z13.web.core.windows.net/pr-20951/docs/index.html

unodevops commented:
⚠️⚠️ The build [169966](https://dev.azure.com/uno-platform//Uno%20Platform/_build/results?buildId=169966) has failed on Uno.UI - CI.

carldebilly committed:
test(ci): improve Uno template install robustness for telemetry Hot Reload tests

- Add best-effort Uno template installation in test infrastructure for CI and local runs
- Log warnings on install failure, non-blocking for local runs
- Refactor test base and initializers to ensure template presence and consistent logging
- Improve diagnostics for telemetry file issues in test setup
- Increases CI reliability for Hot Reload telemetry tests, with resilient local fallback

carldebilly committed:
refactor(test): unify process launching in SolutionHelper via ProcessUtil

- Add ProcessUtil.RunProcessAsync for process execution and output capture
- Refactor SolutionHelper to use ProcessUtil

carldebilly committed:
chore: Fixed code styling issues

unodevops commented:
⚠️⚠️ The build [169984](https://dev.azure.com/uno-platform//Uno%20Platform/_build/results?buildId=169984) has failed on Uno.UI - CI.

carldebilly commented on src/Uno.UI.RemoteControl.Host/Program.cs:
```diff
@@ -1,112 +1,265 @@
+				var startupProperties = new Dictionary<string, string>
+				{
+					["HasSolution"] = (solution != null).ToString(),
+					["MachineName"] = TelemetryHashHelper.Hash(Environment.MachineName),
+					["OSVersion"] = Environment.OSVersion.ToString(),
```
Probably already in the telemetry service, would be duplicate information.

carldebilly committed:
chore: Use the environment variable "UNO_PLATFORM_TELEMETRY_FILE" instead

carldebilly committed:
ci: Trying to fix the build

carldebilly committed:
ci: Trying to fix build

unodevops commented:
🤖 Your Docs stage site is ready! Visit it here: https://unodocsprstaging.z13.web.core.windows.net/pr-20951/docs/index.html

unodevops commented:
🤖 Your WebAssembly Skia Sample App stage site is ready! Visit it here: https://unowasmprstaging.z20.web.core.windows.net/pr-20951/wasm-skia-net9/index.html

carldebilly committed:
ci: improve template install reliability and diagnostics in test infra

carldebilly committed:
docs: update Telemetry.md for processor and startup event changes

unodevops commented:
🤖 Your WebAssembly Skia Sample App stage site is ready! Visit it here: https://unowasmprstaging.z20.web.core.windows.net/pr-20951/wasm-skia-net9/index.html

carldebilly committed:
feat(telemetry): add EventsPrefix support to FileTelemetry and propagate from ServiceCollectionExtensions

unodevops commented:
⚠️⚠️ The build [170250](https://dev.azure.com/uno-platform//Uno%20Platform/_build/results?buildId=170250) has failed on Uno.UI - CI.

carldebilly committed:
test(telemetry): validate ITelemetry injection in dynamic processors and fix formatting issues

carldebilly committed:
ci: Fix build

carldebilly committed:
ci: Trying to fix build

carldebilly committed:
ci: Fix build

unodevops commented:
🤖 Your WebAssembly Skia Sample App stage site is ready! Visit it here: https://unowasmprstaging.z20.web.core.windows.net/pr-20951/wasm-skia-net9/index.html

unodevops commented:
⚠️⚠️ The build [170362](https://dev.azure.com/uno-platform//Uno%20Platform/_build/results?buildId=170362) has failed on Uno.UI - CI.

dr1rrb commented on src/Uno.UI.RemoteControl.Host/Extensibility/AddInsExtensions.cs:
```diff
@@ -1,17 +1,27 @@
+		return builder.ConfigureServices(services =>
+		{
+			// Get telemetry service for add-in discovery and loading
+			var serviceProvider = services.BuildServiceProvider();
```
Is this a recommended pattern to use MS DI? Sounds weird to me to resolve a service during the registration phase.

carldebilly committed:
feat(telemetry): Added a graceful shutdown when parent process dies

carldebilly committed:
feat(ui-tests): Added console output detail to nunit file format

carldebilly committed:
chore: CodeReview / Removed "LoadedAssemblies" from measurements

carldebilly commented on src/Uno.UI.RemoteControl.Host/Extensibility/AddIns.cs:
```diff
@@ -8,90 +9,151 @@
-					result = ProcessHelper.RunProcess("dotnet", DumpTFM("diagnostic"), wd);
+			// Note: We include the targets "on the fly" so if a project uses Microsoft.NET.Sdk instead of Uno.Sdk, we will still have the targets defined.
+			var targetsFile = Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location)!,
+				"DevServer.Custom.Targets");
```
Resharper reformatting... the indentation is correct, it's just how I'm doing it usually.

carldebilly committed:
chore: Applied PR review comments

carldebilly committed:
fix(DevServer): Remove BuildServiceProvider anti-pattern in AddInsExtensions

unodevops commented:
🤖 Your Docs stage site is ready! Visit it here: https://unodocsprstaging.z13.web.core.windows.net/pr-20951/docs/index.html

carldebilly committed:
ci: Fixed build by registering required services

carldebilly committed:
chore: Reduced the verbosity of telemetry (removed useless information)

unodevops commented:
🤖 Your Docs stage site is ready! Visit it here: https://unodocsprstaging.z13.web.core.windows.net/pr-20951/docs/index.html

unodevops commented:
⚠️⚠️ The build [170461](https://dev.azure.com/uno-platform//Uno%20Platform/_build/results?buildId=170461) has failed on Uno.UI - CI.

unodevops commented:
🤖 Your WebAssembly Skia Sample App stage site is ready! Visit it here: https://unowasmprstaging.z20.web.core.windows.net/pr-20951/wasm-skia-net9/index.html

carldebilly committed:
ci(telemetry): resolve DI failures caused by AssemblyLoadContext type duplication

Root cause: dynamically loaded processors/DLLs were loaded in a different AssemblyLoadContext than the host, causing type identity issues and breaking generic DI resolution (e.g. ITelemetry<T>). Solution: ensure all shared types are loaded in the same context. Cleaned up code to rely on generic registration only, no hacks or fallbacks.

unodevops commented:
⚠️⚠️ The build [170559](https://dev.azure.com/uno-platform///_build/results?buildId=170559) has failed on Uno.UI - docs.

carldebilly committed:
chore: Removed the "SolutionId" from telemetry properties (as requested by PR review) and replaced it by _currentPath_ in DevTools.Telemetry

carldebilly committed:
chore: Adjust telemetry to remove sensitive information already handled by the Telemetry package

unodevops commented:
🤖 Your WebAssembly Skia Sample App stage site is ready! Visit it here: https://unowasmprstaging.z20.web.core.windows.net/pr-20951/wasm-skia-net9/index.html

carldebilly committed:
chore: Applied prefixes to telemetry properties and measurements

carldebilly commented on src/Uno.UI.RemoteControl.Host/Program.cs:
```diff
@@ -1,112 +1,255 @@
+			using var cancellationTokenSource = new CancellationTokenSource();
+			var shutdownRequested = false;
+
+			Console.CancelKeyPress += (sender, e) =>
```
That's a best-effort and there's a fallback for automations. It works for unit tests and that's enough for me.

unodevops commented:
🤖 Your WebAssembly Skia Sample App stage site is ready! Visit it here: https://unowasmprstaging.z20.web.core.windows.net/pr-20951/wasm-skia-net9/index.html
